pkg/multiboot/header.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package multiboot
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"io"
+)
+
+var ErrHeaderNotFound = errors.New("Multiboot header not found")
+var ErrFlagsNotSupported = errors.New("Multiboot header flags not supported yet")
+
+const headerMagic = 0x1BADB002
+
+const (
+	FlagHeaderPageAlign  uint32 = 0x00000001
+	FlagHeaderMemoryInfo        = 0x00000002
+
+	// unsupported flags
+	FlagHeaderMultibootVideoMode = 0x00000004
+	FlagHeaderAoutKludge         = 0x00010000
+
+	FlagHeaderUnsupported = 0x0000FFFC
+)
+
+var sizeofHeader = binary.Size(Header{})
+
+type Header struct {
+	Magic    uint32
+	Flags    uint32
+	Checksum uint32
+
+	HeaderAddr  uint32
+	LoadAddr    uint32
+	LoadEndAddr uint32
+	BSSEndAddr  uint32
+	EntryAddr   uint32
+
+	ModeType uint32
+	Width    uint32
+	Height   uint32
+	Depth    uint32
+}
+
+func parseHeader(r io.Reader) (Header, error) {
+	var hdr Header
+	// The Multiboot header must be contained completely within the
+	// first 8192 bytes of the OS image.
```

Comment:
If the non-required fields are not used, could the header be smaller and start closer to the 8192 boundary?

Reply:
Acually, I don't know, code of grub2 and kexec-tools suggest that it could.

Reply:
updated code and tests.